src/ripple/json/MultivarJson.h

```diff
@@ -75,75 +107,222 @@ struct MultivarJson
+            std::integral_constant<unsigned int, Version> const version,
+            Fn fn,
+            Args&&... args) const
+            -> std::invoke_result_t<
+                Fn,
+                Json::Value&,
+                std::integral_constant<unsigned int, Version>,
+                Args&&...> requires requires()
+        {
+            fn(json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(
+                json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
 
-/*
+        // Mutable Json only, integral_constant version
+        template <unsigned int Version, typename Fn>
+        auto
+        operator()(
+            MultivarJson& json,
+            std::integral_constant<unsigned int, Version> const,
+            Fn fn) const
+            -> std::invoke_result_t<Fn, Json::Value&> requires requires()
+        {
+            fn(json.val[index(Version)]);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(json.val[index(Version)]);
+        }
 
-NOTE:
+        // Immutable Json, integral_constant version
+        template <unsigned int Version, typename... Args, typename Fn>
+        auto
+        operator()(
+            MultivarJson const& json,
+            std::integral_constant<unsigned int, Version> const version,
+            Fn fn,
+            Args&&... args) const
+            -> std::invoke_result_t<
+                Fn,
+                Json::Value const&,
+                std::integral_constant<unsigned int, Version>,
+                Args&&...> requires requires()
+        {
+            fn(json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(
+                json.val[index(Version)], version, std::forward<Args>(args)...);
+        }
 
-If a future API version change adds another possible format, change the size of
-`MultiApiJson`, and update `apiVersionSelector()` to return the appropriate
-selection value for the new `apiVersion` and higher.
+        // Immutable Json only, integral_constant version
+        template <unsigned int Version, typename Fn>
+        auto
+        operator()(
+            MultivarJson const& json,
+            std::integral_constant<unsigned int, Version> const,
+            Fn fn) const
+            -> std::invoke_result_t<Fn, Json::Value const&> requires requires()
+        {
+            fn(json.val[index(Version)]);
+        }
+        {
+            static_assert(
+                valid(Version) && index(Version) >= 0 && index(Version) < size);
+            return fn(json.val[index(Version)]);
+        }
```

Comment:
Can we collapse this definition with the above two by defining a custom invoker that does the right thing depending on the signature of `fn`?

Reply:
I am not sure that's a good idea, complexity wise. It would save us 2 overloads (so we get to 2 total, one for integral_constant and another for other types) but at the cost of adding 2 extra overloads elsewhere (or an `if constexpr`) for the invoker and extra glue to make it work.